src/toxcore/dht/server.rs

```diff
@@ -170,70 +186,54 @@ impl Server {
         match packet {
             DhtPacket::PingRequest(packet) => {
                 debug!("Received ping request");
-                let client = self.create_client(&addr, packet.pk);
-                self.handle_ping_req(client, packet)
+                self.handle_ping_req(packet, addr)
             },
             DhtPacket::PingResponse(packet) => {
                 debug!("Received ping response");
-                if let Some(client) = self.get_client(&packet.pk) {
-                    self.handle_ping_resp(client, packet)
-                } else { // If there doesn't exist client in hash, then the PingRequest sent is not from me. Do nothing.
-                    Box::new(future::ok(()))
-                }
+                self.handle_ping_resp(packet)
             },
             DhtPacket::NodesRequest(packet) => {
                 debug!("Received NodesRequest");
-                let client = self.create_client(&addr, packet.pk);
-                self.handle_nodes_req(client, packet)
+                self.handle_nodes_req(packet, addr)
             },
             DhtPacket::NodesResponse(packet) => {
                 debug!("Received NodesResponse");
-                if let Some(client) = self.get_client(&packet.pk) {
-                    self.handle_nodes_resp(client, packet)
-                } else { // If there doesn't exist client in hash, then the NodesRequest sent is not from me. Do nothing.
-                    Box::new(future::ok(()))
-                }
+                self.handle_nodes_resp(packet)
             },
-            DhtPacket::DhtRequest(dr) => {
+            DhtPacket::DhtRequest(packet) => {
                 // The packet kind of DhtRequest is in encrypted payload,
                 // so decrypting is needed first.
-                let payload = dr.get_payload(&self.sk)
-                    .map(|p| p)
-                    .map_err(|e| {
-                        e
-                    });
+                let payload = packet.get_payload(&self.sk);
+                let payload = match payload {
+                    Err(_) => {
+                        return Box::new( future::err(
+                            Error::new(ErrorKind::Other,
+                                format!("DhtRequest::get_payload failed with: {:?}",
+                                    payload.unwrap_err())
```

Comment:
No need to use unwrap_err here - you already match payload, so you can use Err(err).